src/references.rs

```diff
@@ -17,3 +24,168 @@ pub fn null_mut<'a, T: 'static>() -> &'a mut T {
 pub fn not_alloc<'a, T: 'static>() -> &'a mut T {
 	crate::transmute(usize::MAX)
 }
+
+/// Easily dereferencable raw pointer. Can be freely moved or copied. Do you really
+/// desire to have such power? If so, do as you wish. You've been warned.
+///
+/// `Ref<T>` is guaranteed to be the same size as `usize`
+#[derive(Copy, PartialEq, Eq, PartialOrd, Ord)]
+#[repr(C)]
+pub struct Ref<T>(usize, PhantomData<Arc<T>>)
+where
+	T: Sized;
+impl<T> Clone for Ref<T> {
+	fn clone(&self) -> Self {
+		Self(self.0, PhantomData)
+	}
+}
+impl<T> Ref<T> {
+	pub fn new(value: T) -> Self {
+		Self(Box::leak(Box::new(value)) as *const _ as usize, PhantomData)
+	}
+
+	pub fn free(self) {
+		let boxx: Box<T> = crate::transmute(self);
+		drop(boxx);
+	}
+
+	pub fn addr(&self) -> usize {
+		self.0
+	}
+
+	pub fn as_ptr(&self) -> *const T {
+		crate::transmute(self.0)
+	}
+
+	pub fn as_ptr_mut(&mut self) -> *mut T {
+		crate::transmute(self.0)
+	}
+
+	pub fn from_ptr(ptr: *const T) -> Self {
+		Self(ptr as usize, PhantomData)
+	}
+
+	pub fn from_ptr_mut(ptr: *mut T) -> Self {
+		Self(ptr as usize, PhantomData)
+	}
+
+	pub fn into_box(&self) -> Box<T> {
+		crate::transmute(self.0)
+	}
+}
+impl<T> Ref<Ref<T>> {
+	pub fn flatten(self) -> Ref<T> {
+		self.deref().clone()
+	}
+}
+impl<T> AsRef<T> for Ref<T> {
+	fn as_ref(&self) -> &T {
+		crate::transmute(self.0)
+	}
+}
+impl<T> AsMut<T> for Ref<T> {
+	fn as_mut(&mut self) -> &mut T {
+		crate::transmute(self.0)
+	}
+}
+impl<T> Deref for Ref<T> {
+	type Target = T;
+
+	fn deref(&self) -> &Self::Target {
+		crate::transmute(self.0)
+	}
+}
+impl<T> DerefMut for Ref<T> {
+	fn deref_mut(&mut self) -> &mut Self::Target {
+		crate::transmute(self.0)
+	}
+}
+impl<T> From<&T> for Ref<T> {
+	fn from(value: &T) -> Self {
+		crate::transmute(value)
+	}
+}
+impl<T> From<Box<T>> for Ref<T> {
+	fn from(value: Box<T>) -> Self {
+		let reff = crate::transmute(&*value);
+		std::mem::forget(value);
+		reff
+	}
+}
+impl<T> From<Rc<T>> for Ref<T> {
+	fn from(value: Rc<T>) -> Self {
+		crate::transmute(&*value)
+	}
+}
+impl<T> From<Arc<T>> for Ref<T> {
+	fn from(value: Arc<T>) -> Self {
+		crate::transmute(&*value)
+	}
+}
+
+/// Brinding https://rust-lang.github.io/rfcs/0809-box-and-in-for-stdlib.html back with
+/// better C emulation.
+///
+/// `Ref<T>` is guaranteed to be the size of a `usize` for all `Sized` types. For non-`Sized`
+/// types we have no current support anyway.
+pub fn new<T>(value: T) -> Ref<T>
+where
+	T: Sized,
+{
+	Ref::new(value)
+}
+
+/// Accompaning `free` function for unnewing your `new`
+///
+/// Memory under passed reference will be freed. Upon freeing, using the same `Ref<T>`
+/// again or calling `free` on it is Undefined Behavior.
+pub fn free<T>(reff: Ref<T>)
+where
+	T: Sized,
+{
+	reff.free()
+}
```

Comment:
Can remove the module-level fns now that they're methods

Reply:
Those methods exist as to be inclusive towards the C/C++ developer community. Those can be removed, but it will make Rust less accessible for that group.

Reply:
Ye makes sense

Reply:
Ah, those have names which cannot be exported

Reply:
Being declared with `no_mangle`, that is

Reply:
hmmm
maybe `#[export(cve_rs_malloc)]`/free?

Reply:
No such attribute

Reply:
My bad, `export_name`: https://doc.rust-lang.org/reference/abi.html#the-export_name-attribute

Reply:
cargo mommy lmfao